plugins/optimization-detective/debug.php

```diff
@@ -0,0 +1,217 @@
+<?php
+/**
+ * Debug helpers used for Optimization Detective.
+ *
+ * @package optimization-detective
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Registers tag visitors.
+ *
+ * @since n.e.x.t
+ *
+ * @param OD_Tag_Visitor_Registry $registry Tag visitor registry.
+ */
+function od_debug_register_tag_visitors( OD_Tag_Visitor_Registry $registry ): void {
+	$debug_visitor = new Optimization_Detective_Debug_Tag_Visitor();
+	$registry->register( 'optimization-detective/debug', $debug_visitor );
+}
+
+add_action( 'od_register_tag_visitors', 'od_debug_register_tag_visitors', PHP_INT_MAX );
+
+
+/**
+ * Filters additional properties for the element item schema for Optimization Detective.
+ *
+ * @since n.e.x.t
+ *
+ * @param array<string, array{type: string}> $additional_properties Additional properties.
+ * @return array<string, array{type: string}> Additional properties.
+ */
+function od_debug_add_inp_schema_properties( array $additional_properties ): array {
+	$additional_properties['inpData'] = array(
+		'description' => __( 'INP metrics', 'optimization-detective' ),
+		'type'        => 'array',
+		// All extended properties must be optional so that URL Metrics are not all immediately invalidated once an extension is deactivated.
+		'required'    => false,
+		'items'       => array(
+			'type'       => 'object',
+			'required'   => true,
+			'properties' => array(
+				'value'             => array(
+					'type'     => 'number',
+					'required' => true,
+				),
+				'rating'            => array(
+					'type'     => 'string',
+					'enum'     => array( 'good', 'needs-improvement', 'poor' ),
+					'required' => true,
+				),
+				'interactionTarget' => array(
+					'type'     => 'string',
+					'required' => true,
+				),
+			),
+		),
+	);
+	return $additional_properties;
+}
+
+add_filter( 'od_url_metric_schema_root_additional_properties', 'od_debug_add_inp_schema_properties' );
+
+/**
+ * Adds a new admin bar menu item for Optimization Detective debug mode.
+ *
+ * @since n.e.x.t
+ *
+ * @param WP_Admin_Bar $wp_admin_bar The WP_Admin_Bar instance, passed by reference.
+ */
+function od_debug_add_admin_bar_menu_item( WP_Admin_Bar &$wp_admin_bar ): void {
+	if ( ! current_user_can( 'customize' ) && ! wp_is_development_mode( 'plugin' ) ) {
+		return;
+	}
+
+	if ( is_admin() ) {
+		return;
+	}
+
+	$wp_admin_bar->add_menu(
+		array(
+			'id'     => 'optimization-detective-debug',
+			'parent' => null,
+			'group'  => null,
+			'title'  => __( 'Optimization Detective', 'optimization-detective' ),
+			'meta'   => array(
+				'onclick' => 'document.body.classList.toggle("od-debug");',
+			),
+		)
+	);
+}
+
+add_action( 'admin_bar_menu', 'od_debug_add_admin_bar_menu_item', 100 );
+
+/**
+ * Adds inline JS & CSS for debugging.
+ */
+function od_debug_add_assets(): void {
+	if ( ! od_can_optimize_response() ) {
+		return;
+	}
```

Comment:
Oh, well, if we want to access the URL Metrics for unauthenticated requests then we just need to unset the `user_logged_in` key from the array returned by `od_get_normalized_query_vars()` below. In fact, we could obtain the URL Metrics from both and combine them into one `OD_URL_Metric_Group_Collection`, perhaps with a sample size of `2 * od_get_url_metrics_breakpoint_sample_size()`. 

The remaining issue is what to do about the XPaths from the unauthenticated URL Metrics not applying to the current authenticated response (e.g. due to the admin bar). I think we may need an alternative to how we currently check if the current XPath via `$processor->get_xpath()` matches an XPath stored in a URL Metric. In particular, consider 

  | Logged-out | Logged-in
-- | -- | --
Screenshot | ![image](https://github.com/user-attachments/assets/e1621e78-8dde-4bdc-b503-59f8f1133838) | ![image](https://github.com/user-attachments/assets/196e9c88-d796-464c-b281-05b16ed44bae)
XPath | `/*[1][self::HTML]/*[2][self::BODY]/*[1][self::DIV]/*[2][self::MAIN]/*[1][self::DIV]/*[3][self::DIV]/*[1][self::FIGURE]/*[1][self::IMG]` | `/*[1][self::HTML]/*[2][self::BODY]/*[2][self::DIV]/*[2][self::MAIN]/*[1][self::DIV]/*[3][self::DIV]/*[1][self::FIGURE]/*[1][self::IMG]`

Note the slight difference in the index of the `DIV` caused by the admin bar:

```diff
- /*[1][self::HTML]/*[2][self::BODY]/*[1][self::DIV]
+ /*[1][self::HTML]/*[2][self::BODY]/*[2][self::DIV]
```

Maybe we should have a helper that strips out the index from elements which appear as direct children of `BODY` given that arbitrary elements are added/removed at `wp_body_open` and `wp_footer`? If this instead became normalized to:

```
/*[1][self::HTML]/*[2][self::BODY]/*[self::DIV]
```

Or more simply:

```
/*[1][self::HTML]/*[2][self::BODY]/DIV
```

Then this would match the `DIV.wp-site-blocks` both for when the user is logged-in and logged-out. This would be particularly effective for block themes which always have this root `DIV.wp-site-blocks` element from `get_the_block_template_html()`. It would be more prone to confusion in classic themes, for example, if there are two root `DIV` elements for both the header and footer. However, in looking at the core themes, every single theme except for Twenty Twenty has a root `DIV#page` wrapper for the page contents after `wp_body_open()`. The one exception is Twenty Twenty but it uses `HEADER`, `MAIN`, and `FOOTER` at the root. So my concerns are likely unfounded and we can just leave the index off of elements which are children of `BODY`.